core/match_pattern.py

```diff
@@ -168,6 +167,7 @@ def match_pattern(s: str, p: str, child_tab: list[Child], lcp_tab: list[int], su
 
 
 def find_pattern(s: str, p: str, suf_tab: list[int], lcp_tab: list[int]) -> MatchResult:
+    s += '$'
```

Comment:
It's okay to use $ as the sentinel here, because you know that the strings are over {A,C,G,T} and $ is smaller than all those, but in general it is a bit risky. You don't generally know if $ is in the string, nor do you know if it is the smallest value you there (which you might use in some of the algorithms). But as I said, for the purpose of this, it is fine.